File: 2022/2022/day2/day2.1.swift
//
//  day2.1.swift
//  2022
//
//  Created by Teboho Mphure on 2022/12/02.
//

import Foundation


class Day2_1 {
    var myPlays = ["X": 1, "Y": 2, "Z": 3]
    var total = 0
    
    func GetStrategyValue() {
        let lines = FileUtils.ReadFile(directory: "repos/advent-of-code/2022/2022/files/day2.txt")
        
        for line in lines {
            let pair = line.components(separatedBy: " ")
            
            if pair.count == 2 {
                var answer = 0
                answer = EvaluateOptions(first: pair[0], second: pair[1])
                total += answer + myPlays[pair[1]]!
//                switch(pair[0]) {
//                case "A":
//                    answer = EvaluateOptions(first: pair[0], second: "Y")
//                    total += answer + myPlays["Y"]!
//                case "B":
//                    answer = EvaluateOptions(first: pair[0], second: "X")
//                    total += answer + myPlays["X"]!
//                default:
//                    answer = EvaluateOptions(first: pair[0], second: "Z")
//                    total += answer + myPlays["Z"]!
//                }
        
                
            }
        }
        
        print(total)
    }
    
    func EvaluateOptions(first: String, second: String) -> Int {
        var result = 0
        
        if first == "A" { // rock (opp)
            switch second {
            case "X": // rock
                result = 3
            case "Y": // paper
                result = 6
            default: // scissors
                result = 0
            }
        } else if first == "B" { // paper (opp)
            switch second {
            case "X": // rock
                result = 6
            case "Y": // paper
                result = 3
            default: // scissors
                result = 0
            }
        } else { // scissor
            switch second {
            case "X": // rock
                result = 6
            case "Y": // paper
                result = 0
            default: // scissor
                result = 3
            }
        }
        
        return result
    }
}

File: 2024/day2.1.dart
import 'util.dart';

void main() {
  var fileContent = readFile('./files/day2.txt');
  var lines = fileContent.split('\n');
  var total = 0;

  for (final line in lines) {
    var safe = true;
    var positive = false;

    var items = line.split(' ');

    for (var i = 0; i < items.length - 1; i++) {
      var answer = int.parse(items[i]) - int.parse(items[i + 1]);

      if (i == 0) {
        positive = answer > 0 || answer == 0;
      } else if (positive && answer < 0) {
        safe = false;
        break;
      } else if (!positive && answer >= 0) {
        safe = false;
        break;
      }

      if (positive && (answer < 1 || answer > 3)) {
        safe = false;
        break;
      }

      if (!positive && (answer.abs() < 1 || answer.abs() > 3)) {
        safe = false;
        break;
      }
    }
    if (safe) total++;
  }

  print(total);
}
